Add a dummy module for initial state radiation

chanwook-park opened the pull request:
Hi Joern, I added a dummy initial state radiation module that performs

1. taking two (space-like) incoming hard partons that experience hard scattering from `PythiaGun`.
2. initial state radiation from the time when the hard scattering occurs, tau=tau_0 (but currently at t=0), in the backward direction in time until all space-like parton's pt goes below certain pT (or virtuality), creating time-like partons.
3. ordinary shower (MATTER) for time-like partons until tau_0 and push back all resulting partons to `HardProcess` module using `AddParton()`.

These functionalities are all tentative and could be changed as we discuss further.

To use this new module, please use `jetscape->Add(isr);` in PythiaBrickTest.cc and remove `pythiaGun`. The interface between this module and `JetEnergyLoss` module is in line 268-269 in InitialStateRadiation.cc. 
Please check the code and let me know if you have any comments. Thanks a lot!

jhputschke commented:
Hi Chanwook,

a generic question concerning the ISR module. From your implementation you are not planning on using the clock implementation, rather than executing the ISR module per event with an "internal" clock. I think this is fine if we don't expect any further interactions with other modules at this stage of the collision evolution. Maybe you can confirm if this is what is expected and/or if it would be preferable to use the clock implementation.
Furthermore it would be probably good to discuss in more detail if the current setup, inheriting from the HardProcess class is the best approach. In my view it might become a bit more difficult in the future since you basically have to include the hard process generation and the ISR in one class, which would make it more difficult if you for example want to switch from Pythia to a different hard process generator, which might not happen anytime soon, but generically it would be good to take this into consideration. For this pull request I think we can ignore that aspect, to keep going with a pragmatic first test implementation, but certainly one should discuss this further (maybe in one of the next phone meetings, since it might be more efficient to discuss this initially in person).
With that in mind I think it might be good to change the filenames to something including a clear indicating that this is a test, since the overall inheritance dependence might change so maybe InitialStateRadiation --> InitialStateRadiationTest also for the class name. I feel it would be good since it is a good reminder that while fine for first testing we have to come back and think a bit about it.
On a more practical point, I think it would be good if you could convert all "cout" statements to using the VERBOSE or DEBUG features of the JetScapeLogger class. This will make it easy to suppress output if no debug/verbose information is required. 

Thanks

Joern

chanwook-park commented:
Hi Joern. Thank you for nice feedbacks! 

Regarding the clock implementation, I think the best approach would be that the framework provides this ISR module with clock and time step and this module runs within this clock time. The internal time loop and variables are for testing purpose and this could be modified later on if the framework has backward clock implementation. However, if it turns out that this is not efficient, then it would be okay to keep the internal clock in the ISR module.

Also I agree that this module is for 'initial state radiation' not for 'hard processes'. I think this module is rather a special type of energy loss module. Thus, for the partons from the hard process, I think it would be good to call pythia to produce hard partons with initial state radiation off and let this module to do ISR. At this moment, I didn't know how to implement pythia+ISR module within XSCAPE. It would be great to have a in-depth discussion in our meeting.

For the last two comments, I will modify filename/class name/VERBOSE accordingly and make another commit.

jhputschke commented:
Sounds good, let’s discuss further, but with your new commit we can include it to move ahead for testing purposes!

chanwook-park commented:
Hi Joern, I added another commit, where I changed the class/filename from InitialStateRadiation to InitialStateRadiationTest, and replace 'cout' with 'JSDEBUG' as you suggested. Please check out the code and let me know if you have comments.

Thanks!

Chanwook

jhputschke commented:
Looks good.
